### PR TITLE
Implemented applying dropout at inference time

### PR DESCRIPTION
**Pull request by mfomicheva**
Summary:

Implemented Monte Carlo dropout. Added README to reproduce the results from our paper
that applies this idea for unsupervised quality estimation of NMT (joint work of Facebook AI and the University of Sheffield):

Marina Fomicheva, Shuo Sun, Lisa Yankovskaya, Frédéric Blain, Francisco Guzmán, Mark Fishel, Nikolaos Aletras, Vishrav Chaudhary, Lucia Specia. Unsupervised Quality Estimation for Neural Machine Translation. Accepted to TACL

Retaining dropout at test time is not possible in the current code base. The statement
```
if not self.retain_dropout:
  model.eval()
```
in `SequenceGenerator` does not have any effect, since model `training` attribute is already set to False by the method `make_generate_fast_`, which is applied before initializing `SequenceGenerator` in `generate.py`. `make_generate_fast_` throws an exception when trying to set `training` to True after its application. Also, if I am not mistaken `self.training=True` can have other effects, so setting it to True only for the purpose of retaining dropout at test time might be confusing. I propose an alternative implementation where `retain_dropout` is an attribute of FairseqModel class.

# Before submitting

- [N] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [Y] Did you read the [contributor guideline](https://github.com/pytorch/fairseq/blob/master/CONTRIBUTING.md)?
- [Y] Did you make sure to update the docs?   
- [Y] Did you write any new necessary tests?  

## What does this PR do?
New feature.

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by mfomicheva** [comment]
> Thanks for the PR, looks very nice! I left a few comments inline below.
> 
> Also, I'm a bit uneasy about adding a `set_inference_dropout` abstraction, since it doesn't seem that flexible. Another option would be to switch to Dropout modules instead of the functional forms and have each model add its own `--retain-dropout` flag. Then this functionality would be entirely driven by the model and not depend on the SequenceGenerator.
> 
> Something like:
> 
> ```python
> class FairseqDropout(nn.Module):
> 
>     def __init__(self, p, apply_during_inference=False):
>         super().__init__()
>         self.p = p
>         self.apply_during_inference = apply_during_inference
> 
>     def forward(self, x):
>         if self.training or self.apply_during_inference:
>             return F.dropout(x, p=self.p, training=True)
>         else:
>             return x
> ```
> 
> What do you think?

Thank you! I like the idea of introducing a Dropout module and using it instead of F.dropout. This would mean changing all the occurrences of F.dropout() for all the models/modules, right? Just wanted to confirm before I do that.

File: examples/unsupervised_quality_estimation/README.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
+# Unsupervised Quality Estimation for Neural Machine Translation (Fomicheva et al., 2020)
+
+This page includes instructions for reproducing results from the paper Unsupervised Quality Estimation for Neural
+Machine Translation (Fomicheva et al., 2020).
```
[the remark this answers is not in the document]
Comment (mfomicheva) [reply examples/unsupervised_quality_estimation/README.md:4]: Yes, that would be great! I am putting the paper on arxiv now, so I can add the link when it is available.

File: fairseq/options.py [hunk 2]
```diff
@@ -591,6 +591,10 @@ def add_generation_args(parser):
                        help='if set, the last checkpoint are assumed to be a reranker to rescore the translations'),
     group.add_argument('--retain-iter-history', action='store_true',
                        help='if set, decoding returns the whole history of iterative refinement')
+    group.add_argument('--retain-dropout', action='store_true',
+                       help='Use dropout at inference time')
+    group.add_argument('--retain-dropout-modules', default=None, nargs='+', type=str,
+                       help='Use dropout at inference time only for specific modules.')
```
[the remark this answers is not in the document]
Comment (mfomicheva) [reply fairseq/options.py:597]: Yes, as it is defined now, it is applied to all modules by default. But the user might want to avoid applying it to some specific modules. E.g. we didn't apply it to the attention modules in our experiments. Maybe it would make more sense if I change the option `--retain-dropout-modules` to `--exclude-dropout-modules` or something like that?

File: fairseq_cli/generate.py [hunk 3]
```diff
@@ -51,6 +53,11 @@ def _main(args, output_file):
         args.max_tokens = 12000
     logger.info(args)
 
+    # Fix seed for stochastic decoding
+    if args.seed is not None:
+        np.random.seed(args.seed)
+        torch.manual_seed(args.seed)
```
[the remark this answers is not in the document]
Comment (mfomicheva) [reply fairseq_cli/generate.py:59]: I think I would go for Option 2.

**Comment by myleott** [comment]
> Thank you! I like the idea of introducing a Dropout module and using it instead of F.dropout. This would mean changing all the occurrences of F.dropout() for all the models/modules, right? Just wanted to confirm before I do that.

Yep, exactly!

**Comment by mfomicheva** [comment]
Hi Myle,
I have made the required changes. There a lot of differences but most of them are related to passing from `F.dropout()` to having a `FairseqDropout` module.

The tests `test_export` and `test_lstm_jitable` fail when passing `inplace=False` to the `forward` method in the `FairseqDropout` module. I am not sure why exactly it happens, unfortunately.

Also, please let me know if you'd like me to rebase on the most recent changes from master.

**Comment by myleott** [comment]
Sorry about the delay on this! Working on getting it in now

File: examples/unsupervised_quality_estimation/README.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+fairseq-preprocess --srcdict ${MODEL_DIR}/dict.${SRC_LANG}.txt $TGT_DIC --source-lang ${SRC_LANG} 
+--target-lang ${TGT_LANG} --testpref ${TMP}/repeated --destdir ${TMP}/bin-repeated
+```
+
+Produce model scores for the generated translations using `--retain-dropout` option to apply dropout at inference time:
+
+```
+CUDA_VISIBLE_DEVICES=${GPU} fairseq-generate ${TMP}/bin-repeated --path ${MODEL_DIR}/${LP}.pt --beam 5
+ --source-lang $SRC_LANG --target-lang $TGT_LANG --no-progress-bar --unkpen 5 --score-reference --retain-dropout
+ --retain-dropout-modules TransformerModel TransformerEncoder TransformerDecoder TransformerEncoderLayer
```
Comment (myleott) [inline examples/unsupervised_quality_estimation/README.md:88]: One clarifying question here. This was renamed to `--exclude-dropout-modules`, so should this be `--exclude-dropout-modules MultiheadAttention` throughout the README?

No need for you to update the PR btw, I can make the necessary changes before merging.

Comment (mfomicheva) [reply examples/unsupervised_quality_estimation/README.md:88]: Yes, exactly! Sorry, I forgot to update the README.

**Comment by myleott** [comment]
@mfomicheva sorry for the considerable delay! I had to make some related changes to internal code to make everything work (and refactor your PR a bit). Can you please review #2308 and confirm if it looks okay to you?

**Comment by mfomicheva** [comment]
> @mfomicheva sorry for the considerable delay! I had to make some related changes to internal code to make everything work (and refactor your PR a bit). Can you please review #2308 and confirm if it looks okay to you?

Hi @myleott ! Thank you! I have looked at the changes and also rerun some experiments. Everything looks good and I am getting the same results. I like the idea of `FairseqDropout` module having its `make_generation_fast_` method to set up everything, it makes things much cleaner. Thanks again for merging it!
